### PR TITLE
Initial AIX support with kernel/operatingsystem facts

### PR DESCRIPTION
**Pull request by branan**
This is the beginning of some AIX stuff.

This still includes a bit of a hacky workaround for dlopen flags on AIX, so probably shouldn't be merged just yet.

### CODE REVIEW

File: lib/src/facts/aix/kernel_resolver.cc [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        data result;
+        struct utsname name;
+        if (uname(&name) == -1) {
+            LOG_WARNING("uname failed: %1% (%2%): kernel facts are unavailable.", strerror(errno), errno);
+            return result;
+        }
+
+        result.name = name.sysname;
+        result.release = name.release;
+        result.version = name.version;
```
Comment (branan) [inline lib/src/facts/aix/kernel_resolver.cc]: This only differs from the POSIX resolver by one line. Thoughts on making that an ifdef instead of having an entirely new file just for AIX?

Comment (peterhuene) [inline lib/src/facts/aix/kernel_resolver.cc]: A one line `#ifdef` is fine by me; it doesn't sound like that would grow into #ifdef hell.

**Comment by MikaelSmith** [comment]
Lint is unhappy.

File: lib/src/facts/aix/operating_system_resolver.cc [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    operating_system_resolver::data operating_system_resolver::collect_data(collection& facts)
+    {
+        // Default to the base implementation
+        auto result = posix::operating_system_resolver::collect_data(facts);
+        result.architecture = getattr("proc0", "type");
+        result.hardware = getattr("sys0", "modelname");
+
+        // Get OS release stuff
+        static boost::regex regexp_release("(\\d+)\\.(\\d+\\.\\d+\\.\\d+)");
```
Comment (whopper) [inline lib/src/facts/aix/operating_system_resolver.cc]: @branan did we decide to just report the `6.1` from the `6.1.7.15` found in `aix_release.level`? If so, we can probably just do something like:

    static boost::regex regexp_release("(\\d+\\.\\d+)\\.\\d+\\.\\d+");
        file::each_line("/usr/lpp/bos/aix_release.level", [&](string& line) {
            string full_release;
            if (re_search(line, regexp_release, &full_release)) {
                result.release = full_release;
                return false;
            }
            return true;
        });
        return result;

No need to concatenate `major + '.' + minor`, as those get parsed in the `parse_release` function below.

File: CMakeLists.txt [hunk 3]
```diff
@@ -32,6 +37,11 @@ if (MINGW)
     # MinGW crashes with large pre-compiled headers; ours definitely exceeds the limit.
     # See http://stackoverflow.com/questions/10841306/cc1plus-exe-crash-when-using-large-precompiled-header-file
     set(PRECOMPILED_HEADERS FALSE)
+elseif(AIX)
+    # On AIX, precompiled headers are causing crazy crazy failures
+    # with things that need to be extern "C". This seems to clear it
+    # all up.
+    set(PRECOMPILED_HEADERS FALSE)
```
Comment (MikaelSmith) [inline CMakeLists.txt]: No complaints here, pre-compiled headers are a hack.

File: lib/src/facts/aix/operating_system_resolver.cc [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+using namespace std;
+using namespace facter::util;
+namespace execution = facter::execution;
+
+string getattr(string object, string field)
+{
+    string result;
+
+    // To get architecture, can/should we use getconf KERNEL_BITMODE, getconf HARDWARE_BITMODE?
```
Comment (MikaelSmith) [inline lib/src/facts/aix/operating_system_resolver.cc]: Get any answers to this?

File: lib/src/facts/aix/operating_system_resolver.cc [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    execution::each_line(
+        "/usr/sbin/lsattr", {"-El", object, "-a", field},
+        [&](string& line) {
+            if (!line.empty()) {
+                vector<string> tokens;
+                istringstream iss(line);
+                // istream_iterator is a great way to tokenize your life
+                copy(istream_iterator<string>(iss),
```
Comment (MikaelSmith) [inline lib/src/facts/aix/operating_system_resolver.cc]: Is this something we should use more generally, and put in a utility? I think other places we just use something like `boost::split(tokens, line, boost::is_space(), boost::token_compress_on)`.

File: lib/src/facts/aix/operating_system_resolver.cc [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+using namespace facter::util;
+namespace execution = facter::execution;
+
+string getattr(string object, string field)
+{
+    string result;
+
+    // To get architecture, can/should we use getconf KERNEL_BITMODE, getconf HARDWARE_BITMODE?
+    execution::each_line(
+        "/usr/sbin/lsattr", {"-El", object, "-a", field},
```
Comment (MikaelSmith) [inline lib/src/facts/aix/operating_system_resolver.cc]: Is it possible to invoke this once to get all the data we need? Shelling out is something we try to minimize.

File: lib/src/facts/aix/operating_system_resolver.cc (hunk 2 above)
Comment (MikaelSmith) [inline lib/src/facts/aix/operating_system_resolver.cc]: And even if not, why do you use a regex that splits it up, just to rejoin them?

File: lib/src/util/posix/dynamic_library.cc [hunk 7]
```diff
@@ -41,7 +41,11 @@ namespace facter { namespace util {
         close();
 
         // Don't actually perform a load to determine if it is already loaded
+#ifdef _AIX
+        _handle = dlopen(name.c_str(), RTLD_LAZY);
```
Comment (MikaelSmith) [inline lib/src/util/posix/dynamic_library.cc]: Why was this necessary?

File: lib/src/facts/aix/operating_system_resolver.cc (hunk 2 above)
Comment (whopper) [inline lib/src/facts/aix/operating_system_resolver.cc]: Heh, the rejoining bit was copied straight out of the Solaris resolver where we had to do that because some versions needed `major + '_u' + minor` and others needed `major + '.' + minor`. Looks like we just forgot to update this to get rid of that.

File: lib/src/util/posix/dynamic_library.cc (hunk 7 above)
Comment (kylog) [inline lib/src/util/posix/dynamic_library.cc]: https://tickets.puppetlabs.com/browse/FACT-891 has backstory.

File: lib/src/facts/aix/operating_system_resolver.cc (hunk 4 above)
Comment (whopper) [inline lib/src/facts/aix/operating_system_resolver.cc]: @MikaelSmith I think the answer is no, looking at what `lsattr` and `getconf` return:
    getconf KERNEL_BITMODE
    64

    getconf HARDWARE_BITMODe
    64

    lsattr -El proc0 -a type
    type PowerPC_POWER7 Processor type False

    lsattr -El sys0 -a modelname
    modelname IBM,8231-E1D Machine name False

Although I can't entirely recall how we're determining the actual architecture with lsattr. If we're reporting POWERPC..., it isn't consistent with other platforms that just return `x86_64` or `i386`

File: lib/src/util/posix/dynamic_library.cc (hunk 7 above)
Comment (MikaelSmith) [inline lib/src/util/posix/dynamic_library.cc]: Thanks, hadn't made the connection.

File: lib/src/facts/aix/operating_system_resolver.cc (hunk 6 above)
Comment (whopper) [inline lib/src/facts/aix/operating_system_resolver.cc]: It looks like it might not be possible, based on the `lsattr` man page. The `-E` option displays effective values for a specific 'Customized Device' object class (specified with `-l`). Since we're checking both `proc0` and `sys0`, I'm not sure if we can get both to return in one call.

Comment (MikaelSmith) [inline lib/src/facts/aix/operating_system_resolver.cc]: Ok.

File: lib/src/util/posix/dynamic_library.cc (hunk 7 above)
Comment (whopper) [inline lib/src/util/posix/dynamic_library.cc]: I believe this was @branan's workaround for FACT-891, and without this, Facter won't compile on AIX. He mentioned earlier that he was debating whether to leave this hack in until we fix FACT-891, so we don't push a commit that doesn't actually build in AIX.

Comment (MikaelSmith) [inline lib/src/util/posix/dynamic_library.cc]: Sounds fine to me; I'd probably prefer a comment explicitly linking it to FACT-891.

Comment (branan) [inline lib/src/util/posix/dynamic_library.cc]: Absolutely. The status of this hack is the main reason this is marked do
not merge.

I'll move it to its own commit and comment that its a dirty hack

File: lib/src/facts/aix/operating_system_resolver.cc (hunk 4 above)
Comment (branan) [inline lib/src/facts/aix/operating_system_resolver.cc]: I think this is consistent-ish. If the goal of the architecture fact is to
be able to say "what binaries can I drop on this system", I think this is a
reasonable value.

Of course, we don't return feature info (such as SSE support) on x86, so
maybe " power " is enough here

File: lib/src/util/posix/dynamic_library.cc (hunk 7 above)
Comment (kylog) [inline lib/src/util/posix/dynamic_library.cc]: +1 on a discrete commit for this workaround. But if it's accompanied by an explanatory comment (something about "Workaround pending resolution of FACT-891"), this seems reasonable enough. 

And then if we end up living with some of these `ifdef`'s long-term for whatever reason, it would be nice to segregate them in some platform driven headers.

File: lib/src/facts/aix/operating_system_resolver.cc (hunk 5 above)
Comment (branan) [inline lib/src/facts/aix/operating_system_resolver.cc]: We should probably do a quick error-check on the output to ensure we got the number of tokens we expected.

I used the istream_iterator because that's how I'm used to tokenizing. boost::split is probably better. I'll fix this up.

**Comment by branan** [comment]
Just pushed up a big commit to refactor the kernel/os facts back into values similar to what Facter 2 returned. It's kind of messy, I might try to refactor a bit before this gets merged.

**Comment by branan** [comment]
I think this is ready for final review now

File: lib/CMakeLists.txt [hunk 8]
```diff
@@ -112,6 +111,13 @@ if (UNIX)
         set(LIBFACTER_STANDARD_SOURCES ${LIBFACTER_STANDARD_SOURCES} "src/util/posix/scoped_bio.cc")
     endif()
 
+    if (NOT AIX)
```
Comment (MikaelSmith) [inline lib/CMakeLists.txt]: So this still works on Solaris and BSD, just not AIX.

File: lib/inc/internal/facts/aix/kernel_resolver.hpp [hunk 9]
```diff
@@ -0,0 +1,25 @@
+/**
+ * @file
+ * Declares the POSIX kernel fact resolver.
```
Comment (MikaelSmith) [inline lib/inc/internal/facts/aix/kernel_resolver.hpp]: Description needs an update.

File: lib/src/facts/aix/kernel_resolver.cc [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        vector<string> versions;
+
+        // We reimplement part of oslevel here. We parse the rml cache
+        // and use that to determine the full AIX system version. If
+        // the rml cache doesn't exist, we just call out to oslevel to
+        // regenerate it (since we're gonna be slow in that case
+        // anyway).
+        bool success;
+        bool tried_oslevel = false;
+        do {
```
Comment (MikaelSmith) [inline lib/src/facts/aix/kernel_resolver.cc]: This is just 3 steps:
* Try to read .oslevel.rml.cache
* If that fails, run oslevel
* Try reading .oslevel.rml.cache again

It seems simpler to just have a function that reads .oslevel.rml.cache and just have the 3 steps explicitly. The loop complicates things.

File: lib/CMakeLists.txt (hunk 8 above)
Comment (branan) [inline lib/CMakeLists.txt]: Yeah. AIX has its own implementation of the kernel resolver that uses `oslevel` instead of uname output :/

File: lib/src/facts/aix/kernel_resolver.cc [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+                std::tie(oslevel_success, out, err) = execute("/usr/bin/oslevel", {"-s"}, 0, { execution_options::trim_output, execution_options::redirect_stderr_to_stdout, execution_options::merge_environment });
+                if (!oslevel_success) {
+                    LOG_WARNING("oslevel failed: %1%: kernel facts are unavailable", out);
+                    return result;
+                }
+                tried_oslevel = true;
+            }
+        } while (!success);
+
+        sort(versions.rbegin(), versions.rend());
```
Comment (MikaelSmith) [inline lib/src/facts/aix/kernel_resolver.cc]: If we just want the largest number, we could keep track of that when reading the file rather than storing all of them and sorting.

File: lib/src/facts/aix/operating_system_resolver.cc (hunk 4 above)
Comment (MikaelSmith) [inline lib/src/facts/aix/operating_system_resolver.cc]: Should still update the comment.

File: lib/src/facts/aix/operating_system_resolver.cc (hunk 5 above)
Comment (MikaelSmith) [inline lib/src/facts/aix/operating_system_resolver.cc]: `boost::split`

File: lib/src/facts/aix/kernel_resolver.cc (hunk 10 above)
Comment (branan) [inline lib/src/facts/aix/kernel_resolver.cc]: Yeah. I think I originally had this doing multiple retries and decided that was stupid :)

File: lib/src/facts/aix/kernel_resolver.cc (hunk 11 above)
Comment (branan) [inline lib/src/facts/aix/kernel_resolver.cc]: This code is from Whopper, and I didn't dig into it in too much depth. It's probably a pretty faithful recreation of the bash, which doesn't have such niceties. I'll massage it a bit more

**Comment by branan** [comment]
@MikaelSmith pulled RML parsing out to a new method and fixed it to not iterate twice.

**Comment by MikaelSmith** [comment]
:+1:
